From 0029f23e7f27edad3800d4b207d65a4279096462 Mon Sep 17 00:00:00 2001
From: Alexey Kodanev <aleksei.kodanev@bell-sw.com>
Date: Tue, 1 Mar 2022 16:39:21 +0300
Subject: [PATCH] commit: make the installing of alternative packages safer

Move the removal of old alternative packages to the end and
exit at the first error when installing a new alternative
package.
---
 src/commit.c | 29 +++++++++++++++++++++++++++++
 1 file changed, 29 insertions(+)

diff --git a/src/commit.c b/src/commit.c
index 79438db..80ee049 100644
--- a/src/commit.c
+++ b/src/commit.c
@@ -261,6 +261,14 @@ static int run_commit_hooks(struct apk_database *db, int type)
 				    run_commit_hook, &hook);
 }
 
+static void swap_pkg_change(struct apk_change_array **changes, struct apk_change *change)
+{
+	struct apk_change *new_change = apk_change_array_add(changes);
+
+	*new_change = *change;
+	change->old_pkg = NULL;
+}
+
 int apk_solver_commit_changeset(struct apk_database *db,
 				struct apk_changeset *changeset,
 				struct apk_dependency_array *world)
@@ -342,8 +350,27 @@ int apk_solver_commit_changeset(struct apk_database *db,
 	if (run_commit_hooks(db, PRE_COMMIT_HOOK) == -2)
 		return -1;
 
+	if (changeset->num_remove) {
+		struct apk_change_array *remove_changes = NULL;
+
+		foreach_array_item(change, changeset->changes) {
+			if (change->new_pkg || !change->old_pkg || !change->old_pkg->alternative)
+				continue;
+			if (!remove_changes)
+				apk_change_array_init(&remove_changes);
+			swap_pkg_change(&remove_changes, change);
+		}
+		if (remove_changes) {
+			foreach_array_item(change, remove_changes)
+				swap_pkg_change(&changeset->changes, change);
+			apk_change_array_free(&remove_changes);
+		}
+	}
+
 	/* Go through changes */
 	foreach_array_item(change, changeset->changes) {
+		if (!change->old_pkg && !change->new_pkg)
+			continue;
 		r = change->old_pkg &&
 			(change->old_pkg->ipkg->broken_files ||
 			 change->old_pkg->ipkg->broken_script);
@@ -356,6 +383,8 @@ int apk_solver_commit_changeset(struct apk_database *db,
 			     (change->reinstall && pkg_available(db, change->new_pkg)))) {
 				r = apk_db_install_pkg(db, change->old_pkg, change->new_pkg,
 						       progress_cb, &prog) != 0;
+				if (r && change->new_pkg && change->new_pkg->alternative)
+					return -1;
 			}
 			if (r == 0 && change->new_pkg && change->new_pkg->ipkg)
 				change->new_pkg->ipkg->repository_tag = change->new_repository_tag;
-- 
2.39.2

